### PR TITLE
Interpolate particle data

### LINKED ISSUE #566: Interpolate particle properties
**Describe the feature**
Interpolate particle properties from a given set of neighbour particles

**Methods**

- Moving Least Squares polynomial interplation
- Radial basis function

### PR DESCRIPTION
**Pull request by srhgk2**
resolves #566 #570 

Implementation: open for discussion

**Create and insert a new particle**
- a new particle is created from given coordinates in a given cell
- material history is cloned from the nearest neighbour particle
- a linear search function to find the nearest neighbour particle

**Interpolate properties from neighbour particles**
- Create interpolation methods (MOving least squares, radial basis function, etc)
- Pass the interpolation method to particle function which interpolate properties

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -108,7 +108,6 @@ endif()
 
 
 # KaHIP
-if (MPI_FOUND)
```
Comment (kks32) [inline CMakeLists.txt]: could we undo this change please?

File: include/particle.tcc [hunk 2]
```diff
@@ -242,6 +242,137 @@ void mpm::Particle<Tdim>::initialise() {
...
+  }
+  return status;
+}
+
+// Interpolate properties from neighbour particles
+// TODO: how to pass the interpolator (can be MLS or radial basis function)
+// TODO: check other properties
+// TODO: check exception handling
+template <unsigned Tdim>
+bool mpm::Particle<Tdim>::interpolate_neighbour_properties(
```
Comment (kks32) [inline include/particle.tcc]: Shouldn't we be doing this at a mesh level?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/particle.tcc (hunk 2 above)
Comment (bodhinandach) [reply include/particle.tcc]: I would agree to do it from mesh, as particles dont have pointer to its neighbour

File: include/math/mls_polyinterpolation.tcc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  // Iterate over each data point and compute the associated weight
+  for (const auto& pdata : data_points) {
+    // Distance to data point
+    VectorDim distance;
+    for (unsigned dim = 0; dim < Tdim; ++dim)
+      distance(dim) = std::fabs(pdata(dim) - point(dim)) / span;
+    // Spline weight
+    double weight = std::numeric_limits<double>::max();
+    if (spline_order == 2) weight = this->quadratic_spline_weight(distance);
+    if (spline_order == 3) weight = this->cubic_spline_weight(distance);
```
Comment (bodhinandach) [inline include/math/mls_polyinterpolation.tcc:49]: if the spline order is not available, wouldn't it be better to throw error rather than passing std::numeric_limits<double>::max(), which is wrong, I think.

File: include/math/mls_polyinterpolation.tcc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    throw std::runtime_error("MLS Data points are invalid");
+
+  // B Matrix
+  Eigen::Matrix<double, Tnmonomials, 1> B_matrix;
+  B_matrix.setZero();
+  for (unsigned i = 0; i < data_values.size(); ++i)
+    B_matrix += data_values[i] * B_vector_[i];
+
+  // Compute interpolated value at the point of interest
+  pvalue = point_monomials_.transpose() * M_matrix_.inverse() * B_matrix;
```
Comment (bodhinandach) [inline include/math/mls_polyinterpolation.tcc:146]: M_matrix_ is inversed multiple time as you are calling interpolate_polynomial, wouldn't it be better to store M_matrix_.inverse() as I see that M_matrix_ is constant as it is initialized.

File: include/math/mls_polyinterpolation.tcc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  unsigned p = 0;
+  for (const auto& pdata_coord : data_points) {
+    // Monomials
+    Eigen::VectorXd monomials =
+        mpm::Polynomial::evaluate_monomials<Tdim>(poly_order, pdata_coord);
+    // Compute M matrix
+    this->M_matrix_ += this->weights_[p] * monomials * monomials.transpose();
+    ++p;
+  }
+  if (!std::fabs(M_matrix_.determinant()) > 0.)
```
Comment (bodhinandach) [inline include/math/mls_polyinterpolation.tcc:108]: This will include pass if determinant of M_matrix is very very close to zero.

**Review by bodhinandach — COMMENTED** [review]
Some comments from my side. Are you going to interpolate the state vars too with MLS?

File: include/particle.tcc (hunk 2 above)
Comment (srhgk2) [reply include/particle.tcc]: @kks32 Shall we discuss this part more carefully as I have few things to clarify?

File: CMakeLists.txt (hunk 1 above)
Comment (srhgk2) [reply CMakeLists.txt]: done!

File: include/particle.tcc (hunk 2 above)
Comment (srhgk2) [reply include/particle.tcc]: Moved to mesh.
